staging/src/k8s.io/component-base/metrics/prometheusextension/timing_histogram.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prometheusextension
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"k8s.io/utils/clock"
+)
+
+// GaugeOps is the part of `prometheus.Gauge` that is relevant to
+// instrumented code.
+// This factoring should be in prometheus, analogous to the way
+// it already factors out the Observer interface for histograms and summaries.
+type GaugeOps interface {
+	// Set is the same as Gauge.Set
+	Set(float64)
+	// Inc is the same as Gauge.inc
+	Inc()
+	// Dec is the same as Gauge.Dec
+	Dec()
+	// Add is the same as Gauge.Add
+	Add(float64)
+	// Sub is the same as Gauge.Sub
+	Sub(float64)
+
+	// SetToCurrentTime the same as Gauge.SetToCurrentTime
+	SetToCurrentTime()
+}
+
+// A TimingHistogram tracks how long a `float64` variable spends in
+// ranges defined by buckets.  Time is counted in nanoseconds.  The
+// histogram's sum is the integral over time (in nanoseconds, from
+// creation of the histogram) of the variable's value.
+type TimingHistogram interface {
+	prometheus.Metric
+	prometheus.Collector
+	GaugeOps
+}
+
+// TimingHistogramOpts is the parameters of the TimingHistogram constructor
+type TimingHistogramOpts struct {
+	Namespace   string
+	Subsystem   string
+	Name        string
+	Help        string
+	ConstLabels prometheus.Labels
+
+	// Buckets defines the buckets into which observations are
+	// accumulated. Each element in the slice is the upper
+	// inclusive bound of a bucket. The values must be sorted in
+	// strictly increasing order. There is no need to add a
+	// highest bucket with +Inf bound. The default value is
+	// prometheus.DefBuckets.
+	Buckets []float64
+
+	// The initial value of the variable.
+	InitialValue float64
+}
+
+// NewTimingHistogram creates a new TimingHistogram
+func NewTimingHistogram(opts TimingHistogramOpts) (TimingHistogram, error) {
+	return NewTestableTimingHistogram(clock.RealClock{}, opts)
+}
+
+// NewTestableTimingHistogram creates a TimingHistogram that uses a mockable clock
+func NewTestableTimingHistogram(clock clock.PassiveClock, opts TimingHistogramOpts) (TimingHistogram, error) {
+	desc := prometheus.NewDesc(
+		prometheus.BuildFQName(opts.Namespace, opts.Subsystem, opts.Name),
+		opts.Help,
```

Comment:
@logicalhan suggested offline that we should (given the experimental nature of the new metric type) just prefix it with EXPERIMENTAL METRIC TYPE

[Same in other places below]

Reply:
You mean prefix the `# TYPE` line (how would I accomplish that?) or prefix the `# HELP` line?

Reply:
I'd expose a `NewExperimentalDesc` which wraps and returns from `prometheus.NewDesc` and prepends the prefix in for the Help text.

Reply:
or you can just patch directly here.

Reply:
I think perhaps I understand the issue, and have a suggestion for how to address it.

The issue is, what makes a scrape produce a `# HELP` and a `# TYPE` line that work together to be accurate and make sense?

With no change to the `# TYPE` line --- and I think this is what I want --- the solution involves making the `# HELP` line describe a histogram --- a bunch of observations of some variable.  For a TimingHistogram --- which the instrumented code treats like a Gauge --- the Help field in the Opts would normally be a noun phrase that identifies the variable.  We could make TimingHistogram wrap that to become "Observations from the end of every nanosecond of <noun phrase identifying variable>".

Reply:
No - I think the request it to explicitly add verbatim

"EXPERIMENTAL METRIC TYPE"

to the HELP line of those metrics.

So switch to:
```
fmt.Sprintf("EXPERIMENTAL METRIC TYPE; %s", opts.Help)
```